src/Microsoft.AspNetCore.Mvc.Core/ApplicationModels/InferParameterBindingInfoConvention.cs

```diff
@@ -166,5 +170,12 @@ private bool IsComplexTypeParameter(ParameterModel parameter)
                 .GetMetadataForType(parameter.ParameterInfo.ParameterType);
             return metadata.IsComplexType && !metadata.IsCollectionType;
         }
+
+        private static bool IsFormFile(Type parameterType)
```

Comment:
`BindingSource.Form` is non-greedy, while `BindingSource.FormFile` that's inferred from model metadata is greedy. The difference causes the parameter name to be reset by `ApiController`.